stripe/api_resources/abstract/test_helpers.py

```diff
@@ -3,8 +3,15 @@
 from stripe import error
 from urllib.parse import quote_plus
 
+from typing import TypeVar, ClassVar, Any
+from typing_extensions import Protocol
 
-class APIResourceTestHelpers:
+from stripe.api_resources.abstract.api_resource import APIResource
+
+T = TypeVar("T", bound=APIResource)
+
+
+class APIResourceTestHelpers(Protocol[T]):
```

Comment:
No! Generic[T] works!

Reply:
Oh, no I remembered what this was for.

We want to enforce that everybody who extends from `APIResourceTestHelpers` provides an appropriate `_resource_cls` field. So it is an "interface". If we extend from `Protocol[T]` then you get a type error if you omit `_resource_cls` from the type definition of e.g. `Terminal.Reader.TestHelpers`.

It's not perfect because the implementor of `APIResourceTestHelpers` can still forget to do e.g.
`Reader.TestHelpers._resource_cls = Reader` after the classes have been defined (it's not possible to do this inline because a class variable isn't in scope inside its own definition and TestHelpers is nested inside the containing resource), but I do think it helps clarify the fact that e.g. Terminal.Reader.TestHelpers has the responsibility to set `_resource_cls`.

Reply:
I did end up having to [Fix test_helpers](https://github.com/stripe/stripe-python/pull/1020/commits/b531c68e4ef5af5735a4bc0955639ea8b048c7a9) to actually give the Protocol teeth, though.

Reply:
We could make this an [abstract base class](https://docs.python.org/3/library/abc.html), I guess. That might feel better given that APIResourceTestHelpers provides implementations, too, but if you google "abstract base class vs protocol python" I get the idea that it's kind of six of one, half dozen of the other

Reply:
>We could make this an [abstract base class](https://docs.python.org/3/library/abc.html), I guess. That might feel better given that APIResourceTestHelpers provides implementations, too, but if you google "abstract base class vs protocol python" I get the idea that it's kind of six of one, half dozen of the other

Understood. Thank you for the explanation. I'll leave it to you to pick the option you prefer.